src/silo/query_engine/actions/fasta_aligned.cpp

```diff
@@ -68,74 +72,138 @@ std::string reconstructSequence(
+   std::vector<std::string> aa_sequence_names;
+
+   assortSequenceNamesInto(database, sequence_names, nuc_sequence_names, aa_sequence_names);
+
+   uint32_t partition_index = 0;
+   std::optional<Range<uint32_t>> remaining_result_row_indices{};
+   return QueryResult{[nuc_sequence_names = std::move(nuc_sequence_names),
+                       aa_sequence_names = std::move(aa_sequence_names),
+                       bitmap_filter =
+                          std::make_shared<std::vector<OperatorResult>>(std::move(bitmap_filter)),
+                       remaining_result_row_indices,
+                       &database,
+                       partition_index](std::vector<QueryResultEntry>& results) mutable {
+      while (partition_index < database.partitions.size()) {
+         auto& bitmap = (*bitmap_filter)[partition_index];
+         if (!remaining_result_row_indices) {
+            remaining_result_row_indices = {{0, uint64ToUint32(bitmap->cardinality())}};
          }
-         for (const auto& aa_sequence_name : aa_sequence_names) {
-            const auto& aa_store = database_partition.aa_sequences.at(aa_sequence_name);
-            entry.fields.emplace(
-               aa_sequence_name, reconstructSequence<AminoAcid>(aa_store, sequence_id)
+
+         // The range of results to fully process in this batch
+         Range<uint32_t> result_row_indices =
+            remaining_result_row_indices->take(PARTITION_CHUNK_SIZE);
+         // Remove the same range from the result rows that need to be
+         // created for the current partition
+         remaining_result_row_indices->mutDrop(PARTITION_CHUNK_SIZE);
+
+         if (!result_row_indices.isEmpty()) {
+            SPDLOG_DEBUG(
+               "FastaAligned::execute: refill QueryResult for partition_index {}/{}, {}/{}",
+               partition_index,
+               database.partitions.size(),
+               result_row_indices.toString(),
+               remaining_result_row_indices->beyondLast()
             );
+
+            const auto& database_partition = database.partitions[partition_index];
+            for (const uint32_t row_id : *bitmap) {
+               results.emplace_back(makeEntry(
+                  database.database_config.schema.primary_key,
+                  database_partition,
+                  nuc_sequence_names,
+                  aa_sequence_names,
+                  row_id
+               ));
+
+               result_row_indices.mutRest();
+               if (result_row_indices.isEmpty()) {
+                  // Finished the batch. Remove processed `row_id`s;
+                  // we already removed the corresponding result
+                  // indices from `remaining_result_row_indices`.
+                  bitmap->removeRange(0, inc(row_id));
+                  return;
+               }
+            }
+            std::cerr << "ERROR: ran out of bitmap before finishing result_row_indices\n"
+                      << std::flush;
```

Comment:
Maybe we should have a PANIC macro that takes the argument, and whose implementation can be changed without changing every use, both for logging and for changing from abort to exceptions?

Reply:
Does `abort` have an advantage over exceptions?

Exceptions in general have the advantage that the caller can react and the program doesn't simply terminate without warning. In C++ we don't get stacktraces, so a log statement helps as a workaround there. If we go for exceptions, I think `LOG_AND_THROW` might be good.

Reply:
As I described further above, currently the only way I can get a backtrace is via abort, because we haven't managed to not catch all exceptions. Only uncaught exceptions would get me a backtrace.

My logic is this:

- I don't just want the ability to get backtraces, but also the ability to inspect the context of an uncaught error. That's why I consider libraries that get me a textual backtrace to be pointless: they cost compilation/link time, and only get me the backtrace, not the ability to inspect; neither would it be possible to dump cores this way.

- The debugger is for both of these (gdb). To allow the debugger to work, these kinds of exceptions must terminate the program. But we haven't managed to do this so far. So I use abort as a workaround.

- That's why I suggest a PANIC macro, then once we are able to *not* catch exceptions that represent bugs, we can just change the implementation of the PANIC macro without having to change the usage sites, and we have gotten rid of the use of abort with a single-file change.

Reply:
In principle that sounds reasonable.

Since this is API code, I'd like to avoid `aborts`, so that SILO doesn't fully crash and needs to restart.

Reply:
I'll add the PANIC macro. It will abort if DEBUG env var is set, throw an exception otherwise.